Choices: add `@change-selection` event

hamannjames opened the pull request:
It would be good if the choices component issued an on change event similar to other inputs and selects. I added the capability for the user to use their own @change directive on the component. This is passed into the $attributes iterable and there is an added function called onChangeFunc that will return the value passed. Then, in the html rendered, onChange is set up in x-data as an anonymous function that wraps around the passed value. This protects the other rendered code. When toggle is called, the onchange function is triggered. The anonymous function is passed the value of selection and this is exposed to the user as a $value variable, which they can do cool stuff with, like emit an event or something. This has been tested in both single and multi choice components. It works well and does not break any other functionality according to my tests.

<hr />

![image](https://github.com/robsontenorio/mary/assets/118955/716d302f-ab1e-47ef-8610-ae3bfca77b50)

hamannjames commented:
If you want me to change this implementation, I am happy to change it based on feedback. I really want to contribute to this because I think it is an awesome tool.

robsontenorio commented:
And if you just use “this.$dispatch” instead? 

this.$dispatch(“change”, …)


https://livewire.laravel.com/docs/events#dispatching-livewire-events-from-alpine

robsontenorio commented on src/View/Components/Choices.php:
```diff
@@ -163,10 +154,15 @@ public function render(): View|Closure|string
+                                        target: this.$refs.searchInput,
+                                        value: this.selection
+                                    }
+                                }))
```
The simple one does not work as expected?

this.$dispatch

hamannjames replied:
It does not work quite as expected. You cannot listen for the event on the input itself. You have to listen on the window with something like x-on:change.window. And the reason I wanted to add this feature was for the change event to feel like a normal browser event, like you would listen for with any other input. With this approach, you can do something like @change="console.log(e)" right on the choices element, and you can capture the event and do stuff with it. It is a custom event so it isn't quite the same as a normal browser event, but it works

I think dispatching a livewire event would be great, but it is also an extra feature that I wasn't trying to add with this pull request. I just wanted to add a feature that would mimic a browser event.

robsontenorio commented:
I have tried your implementation on searchable Choices. 

**After search for something there is a warning.**

<img width="927" alt="image" src="https://github.com/robsontenorio/mary/assets/118955/ff77178e-1ba7-45bc-af9e-c6dd9d555780">

hamannjames commented:
Ah, of course, I did not account for the organic change event as the user is typing into the search. Let me look into this more. I obviously didn't test this enough. Thanks for pointing it out.

robsontenorio commented:
Choices is by far the most complex component. I think you should investigate an alternative solution like bubble/catch/bubble approach, or something similar. 

The tricky part is that extra blade  attributes like “@change” are rendered on a different level than toggle list.

hamannjames commented:
Hi Rob. I made another change. Let me know what you think. Appreciate all your feedback so far. 
 
I think this implementation is better. It adds a function that can be called by other methods on the component (in javascript) to dispatch the event in a uniform way. It allows the caller to override the text input value, and it bubbles up. This lets the user treat the event like a normal browser event by calling target.value. 

I think this solution is good because it doesn't override the default change event behavior, but mimics it in a realistic way when the user selects, removes, or clears items. It passes the value in a "normal" browser way by using search input value, but also in a more digestable way via the details parameter if desired. 

I also played around with preventing the regular search change events altogether. You can do it by adding an onchange handler to the search input that prevents events from bubbling up that aren't the custom event. It works, but also didn't feel right to add at this stage just yet.

robsontenorio commented:
Now, taking the same searchable example, it triggers the change event twice.


https://github.com/robsontenorio/mary/assets/118955/288209db-844c-4724-affd-f8b98dbece04

hamannjames commented:
That is expected in its current iteration, although it is not expected that it should log 1 twice.

What is happening is when you search  and then select an option the native change even on the search input is being triggered and the custom change event is being triggered. The native search is being triggered because focus is being taken away from the search to select an element. That is my best guess at least. You can see this by logging this in a change event handler on x-choices:

`@change="console.log($event instanceof CustomEvent)"`

You can see when you search and select, one is false, one is true.

I kept this in because I didn't want to touch the native browser event. However, I can see how it would be confusing for the user. It feels wrong to prevent the native search input event but also wrong to not fire a change event when the user select something. It is also hard to know the best solution because this component behaves like a select and like a text, all in one.

Right now, I have two solutions in mind. One is to block the native search event from happening entirely. You can do this by adding a change handler on the search input that will stop propagation if the event is not my custom event. This will only fire the event on selection and deselection, not on search. 

Another is to only block the search event when the user select something and I am going to reset the input. I haven't pursued this yet, but it would probably entail adding a listener and callback that will only block the next native event, then remove the listener. 

Or, like you said, I could just use a custom event name. That would be a lot simpler, but also is not what I was intending with the feature. I wanted to mimic a regular change event

robsontenorio commented:
It should trigger once after user toggle options on list. Otherwise it cause issues if someone try to use it as `@change="$wire.thing(...)"`

If it simpler to use a custom name like `@change-selection`, let's stick to this approach.

At end of day people expect to know what's is current "selection ID(s)" after toggle items.

hamannjames commented:
Agreed. I changed the event name to "change-selection" and simplified the function. Now someone can listen for the event with something like:

`@change-selection="console.log($event)"`

robsontenorio commented:
We have the ChoicesOffline component , and appreciate  if could replicate this on same PR. Thanks !

hamannjames commented:
Done! tested it on single, multiple, searchable, and non-searchable. I just added the same code I wrote in choices to choicesoffline.

robsontenorio commented:
@hamannjames  It is released! Thanks for contributing !

hamannjames commented:
Yay! @robsontenorio I might add a snippet about the event to the docs and do a PR there. Excited to contribute to this awesome project!

robsontenorio commented:
I did it.

hamannjames commented:
Awesome!

On Mon, Feb 19, 2024 at 1:55 PM Robson Tenório ***@***.***>
wrote:

> I did it.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/robsontenorio/mary/pull/297#issuecomment-1953201012>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AELBSG3HIITC6LSO5XV4DSLYUPC3JAVCNFSM6AAAAABDHXOUDKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNJTGIYDCMBRGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
- James Hamann

Web Developer
United Way of King County
206-496-7825
